specs/networking/rpc-interface.md

```diff
@@ -0,0 +1,292 @@
+The "method ID" fields in the below messages refer to the `method` field in the request structure above.
+
+The first 1,000 values in `response_code` are reserved for system use. The following response codes are predefined:
+
+1. `0`: No error.
+2. `10`: Parse error.
+2. `20`: Invalid request.
+3. `30`: Method not found.
+4. `40`: Server error.
+
+### Alternative for Non-`libp2p` Clients
+
+Since some clients are waiting for `libp2p` implementations in their respective languages. As such, they MAY listen for raw TCP messages on port `9000`. To distinguish RPC messages from other messages on that port, a byte prefix of `ETH` (`0x455448`) MUST be prepended to all messages. This option will be removed once `libp2p` is ready in all supported languages.
+
+## Messages
+
+### Hello
+
+**Method ID:** `0`
+
+**Body**:
+
+```
+(
+    network_id: uint8
+    chain_id: uint8
+    latest_finalized_root: bytes32
+    latest_finalized_epoch: uint64
+    best_root: bytes32
+    best_slot: uint64
+)
+```
+
+Clients exchange `hello` messages upon connection, forming a two-phase handshake. The first message the initiating client sends MUST be the `hello` message. In response, the receiving client MUST respond with its own `hello` message.
+
+Clients SHOULD immediately disconnect from one another following the handshake above under the following conditions:
+
+1. If `network_id` belongs to a different chain, since the client definitionally cannot sync with this client.
+2. If the `latest_finalized_root` shared by the peer is not in the client's chain at the expected epoch. For example, if Peer 1 in the diagram below has `(root, epoch)` of `(A, 5)` and Peer 2 has `(B, 3)`, Peer 1 would disconnect because it knows that `B` is not the root in their chain at epoch 3:
+
+```
+              Root A
+
+              +---+
+              |xxx|  +----+ Epoch 5
+              +-+-+
+                ^
+                |
+              +-+-+
+              |   |  +----+ Epoch 4
+              +-+-+
+Root B          ^
+                |
++---+         +-+-+
+|xxx+<---+--->+   |  +----+ Epoch 3
++---+    |    +---+
+         |
+       +-+-+
+       |   |  +-----------+ Epoch 2
+       +-+-+
+         ^
+         |
+       +-+-+
+       |   |  +-----------+ Epoch 1
+       +---+
+```
+
+Once the handshake completes, the client with the higher `latest_finalized_epoch` or `best_slot` (if the clients have equal `latest_finalized_epoch`s) SHOULD request beacon block roots from its counterparty via `beacon_block_roots` (i.e., RPC method `10`).
+
+### Goodbye
+
+**Method ID:** `1`
+
+**Body:**
+
+```
+(
+    reason: uint64
+)
+```
+
+Client MAY send `goodbye` messages upon disconnection. The reason field MAY be one of the following values:
+
+- `1`: Client shut down.
+- `2`: Irrelevant network.
+- `3`: Fault/error.
+
+Clients MAY define custom goodbye reasons as long as the value is larger than `1000`.
+
+### Get Status
+
+**Method ID:** `2`
+
+**Request Body:**
+
+```
+(
+	sha: bytes32
```

Comment:
I didn't ask when this got drafted, sorry. What is sha here?

Reply:
The commit hash of the node.

Reply:
What is that? Best root?